### PR TITLE
refactor(forestadmin-client): infer raw tree type from plain condition tree

### PR DESCRIPTION
**Pull request by Thenkei**
### General

- [ ] Write an explicit title for the Pull Request, following [Conventional Commits specification](https://www.conventionalcommits.org)
- [ ] Test manually the implemented changes
- [ ] Validate the code quality (indentation, syntax, style, simplicity, readability)

### Security

- [ ] Consider the security impact of the changes made

### CODE REVIEW

File: packages/forestadmin-client/src/permissions/types.ts [hunk 1]
```diff
@@ -1,5 +1,92 @@
...
+
+export const RAW_FILTER_OPERATORS = [
+  ...UNIQUE_OPERATORS,
+  ...INTERVAL_OPERATORS,
+  ...OTHER_OPERATORS,
+];
+
+export type RawOperator = typeof RAW_FILTER_OPERATORS[number];
+
+export const CUSTOM_ACTION_FILTER_LEAF_SOURCES = ['data', 'input'];
```
Comment (Thenkei) [inline packages/forestadmin-client/src/permissions/types.ts:68]: Arf

**Review by ghusse — APPROVED** [review]

File: packages/forestadmin-client/src/permissions/types.ts (hunk 1 above)
Comment (ghusse) [reply packages/forestadmin-client/src/permissions/types.ts:68]: Woopsi

File: packages/forestadmin-client/src/permissions/types.ts [hunk 2]
```diff
@@ -1,5 +1,88 @@
...
+  'previous_year',
+  'previous_month_to_date',
+  'previous_quarter_to_date',
+  'previous_week_to_date',
+  'previous_x_days_to_date',
+  'previous_x_days',
+  'previous_year_to_date',
+] as const;
+
+const OTHER_OPERATORS = [
```
Comment (VincentMolinie) [inline packages/forestadmin-client/src/permissions/types.ts:33]: The thing is you are duplicating code, why not keep using the datasource-toolkit 🤔  ?

Comment (Thenkei) [reply packages/forestadmin-client/src/permissions/types.ts:33]: It's not. This one is snake case (front and back format) datasource-toolkit internal type use UpperPascal case. It's done on purpose to expose a simpler and cleaner interface to users via agent QueryInterface. 💯

File: packages/forestadmin-client/src/permissions/types.ts [hunk 3]
```diff
@@ -1,5 +1,88 @@
...
+  conditions: Array<GenericRawTree<T>>;
+};
+
+type GenericRawTree<T> = GenericRawTreeBranch<T> | T;
+
+export type RawTreeLeaf = {
+  field: string;
+  operator: RawOperator;
+  value?: unknown;
+};
```
Comment (VincentMolinie) [inline packages/forestadmin-client/src/permissions/types.ts:83]: May I suggest:
```js
type CamelToSnakeCase<S extends string> = S extends `${infer T}${infer U}`
  ? `${T extends Capitalize<T> ? '_' : ''}${Lowercase<T>}${CamelToSnakeCase<U>}`
  : S;

type PascalToSnakeCase<S extends string> = S extends `${infer T}${infer U}`
  ? `${Lowercase<T>}${CamelToSnakeCase<U>}`
  : S;

type KeysToSnakeCase<T> = T extends object
  ? {
      [K in keyof T]: T[K] extends string ? PascalToSnakeCase<T[K]> : KeysToSnakeCase<T[K]>;
    }
  : T;

export type GenericRawTreeLeaf = KeysToSnakeCase<PlainConditionTreeLeaf>;
export type GenericRawTreeBranch = KeysToSnakeCase<PlainConditionTreeBranch>;
```

**Pull request merged**